### PR TITLE
feat: premium application subscriptions

### LINKED ISSUE #9856: Add support for entitlement (App Subscriptions)
### Which application or package is this feature request for?

discord.js

### Feature

- https://github.com/discord/discord-api-docs/pull/6451

<hr>

#### New entitlement events:
https://discord.com/developers/docs/monetization/entitlements#gateway-events
> Fires when a user subscribes to a SKU. Contains an entitlement object.

#### PREMIUM_REQUIRED Interaction Response
https://discord.com/developers/docs/monetization/entitlements#using-entitlements-in-interactions
> If your app has monetization enabled, it will have access to a new [PREMIUM_REQUIRED interaction response (type: 10)](https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-response-object-interaction-callback-type). This can be sent in response to any kind of interaction. It does not allow a content field.

#### Entitlements class/endpoints
https://discord.com/developers/docs/monetization/entitlements#entitlement-object
```js
{
    "id": "1019653849998299136",
    "sku_id": "1019475255913222144",
    "application_id": "1019370614521200640",
    "user_id": "771129655544643584",
    "promotion_id": null,
    "type": 8,
    "deleted": false,
    "gift_code_flags": 0,
    "consumed": false,
    "starts_at": "2022-09-14T17:00:18.704163+00:00",
    "ends_at": "2022-10-14T17:00:18.704163+00:00",
    "guild_id": "1015034326372454400",
    "subscription_id": "1019653835926409216"
}
```

### Ideal solution or implementation

```js
client
.on('entitlementCreate', () => {})
.on('entitlementUpdate', () => {})
.on('entitlementDelete', () => {})
```

### Alternative solutions or implementations

_No response_

### Other context

_No response_

### PR DESCRIPTION
**Pull request by almeidx**
**Please describe the changes this PR makes and why it should be merged:**

Adds SKU and Entitlement support to core and discord.js
Resolves #9856

Upstream:
- https://github.com/discord/discord-api-docs/pull/6451
- https://github.com/discord/discord-api-docs/pull/6452
- https://github.com/discord/discord-api-docs/pull/6477
- https://github.com/discord/discord-api-docs/pull/6548

Depends on:
- https://github.com/discordjs/discord-api-types/pull/833

**Status and versioning classification:**

### CODE REVIEW

File: packages/discord.js/src/managers/EntitlementManager.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+   * <warn>If both `before` and `after` are provided, only `before` is respected</warn>
+   */
+
+  /**
+   * Fetches entitlements for this application
+   * @param {FetchEntitlementsOptions} [options={}] Options for fetching the entitlements
+   * @returns {Promise<Collection<Snowflake, Entitlement>>}
+   */
+  async fetch(options = {}) {
+    const query = makeURLSearchParams({
```
Comment (jaw0r3k) [inline packages/discord.js/src/managers/EntitlementManager.js:59]: - https://github.com/discordjs/discord-api-types/pull/833#discussion_r1367797264

Comment (almeidx) [reply packages/discord.js/src/managers/EntitlementManager.js:59]: That comment was made for the List SKUs endpoint. This is for the List Entitlements endpoint

File: packages/discord.js/src/structures/interfaces/InteractionResponses.js [hunk 2]
```diff
@@ -260,6 +260,22 @@ class InteractionResponses {
     this.replied = true;
   }
 
+  /**
+   * Responds to the interaction with an upgrade button.
+   * <warn>Only available for applications with monetization enabled.</warn>
+   * @returns {Promise<void>}
+   */
+  async sendPremiumUpgradeButton() {
```
Comment (jaw0r3k) [inline packages/discord.js/src/structures/interfaces/InteractionResponses.js:268]: Maybe just premiumRequired?

File: pnpm-lock.yaml [hunk 3]
```diff
@@ -661,7 +661,7 @@ importers:
         version: 2.2.2
       discord-api-types:
         specifier: 0.37.60
-        version: 0.37.60
+        version: link:../../../discord-api-types
```
Comment (jaw0r3k) [inline pnpm-lock.yaml:664]: Change it afterwards

File: packages/discord.js/src/structures/interfaces/InteractionResponses.js (hunk 2 above)
Comment (vladfrangu) [inline packages/discord.js/src/structures/interfaces/InteractionResponses.js:268]: I concur, this name is not ideal, especially if Discord ever adds in more options. I'd just name this `sendPremiumRequired`

File: packages/discord.js/src/structures/interfaces/InteractionResponses.js [hunk 4]
```diff
@@ -260,6 +260,22 @@ class InteractionResponses {
...
+    if (this.deferred || this.replied) throw new DiscordjsError(ErrorCodes.InteractionAlreadyReplied);
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/interfaces/InteractionResponses.js:269]: Can you not defer and then send this? really?? or do you need to edit it in the reply?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/interfaces/InteractionResponses.js [hunk 5]
```diff
@@ -260,6 +260,22 @@ class InteractionResponses {
     this.replied = true;
   }
 
+  /**
+   * Responds to the interaction with an upgrade button.
+   * <warn>Only available for applications with monetization enabled.</warn>
```
Comment (Jiralite) [inline packages/discord.js/src/structures/interfaces/InteractionResponses.js:265]: This is not a warning, this is informational.

**Comment by Jiralite** [comment]
Will partials be added here in case? Are they even necessary?

**Review by vladfrangu — DISMISSED** [review]

**Review by Dymidless — APPROVED** [review]

**Comment by almeidx** [comment]
@Jiralite I don't believe so. Discord sends the full entitlement object everywhere

File: packages/discord.js/typings/index.d.ts [hunk 6]
```diff
@@ -3921,6 +3979,37 @@ export class ChannelManager extends CachedManager<Snowflake, Channel, ChannelRes
...
+export interface UserEntitlementCreateOptions {
+  sku: SKUResolvable;
+  user: UserResolvable;
+}
+
+export interface FetchEntitlementOptions {
+  limit?: number;
+  guild?: GuildResolvable;
+  user?: UserResolvable;
+  skus?: SKUResolvable[];
```
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:3999]:
```suggestion
  skus?: readonly SKUResolvable[];
```

**Review by Jiralite — COMMENTED** [review]
Thoughts on the entitlement manager being sweepable?

**Comment by Jiralite** [comment]
Please resolve conflicts!

File: packages/discord.js/src/managers/EntitlementManager.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  /**
+   * Fetches entitlements for this application
+   * @param {FetchEntitlementsOptions} [options={}] Options for fetching the entitlements
+   * @returns {Promise<Collection<Snowflake, Entitlement>>}
+   */
+  async fetch({ limit, guild, user, skus, excludeEnded, cache = true, before, after } = {}) {
+    const query = makeURLSearchParams({
+      limit,
+      guild_id: this.client.guilds.resolveId(guild),
+      user_id: this.client.users.resolveId(user),
```
Comment (vladfrangu) [inline packages/discord.js/src/managers/EntitlementManager.js:63]: Shouldn't this get a `prop && this.client.blah.resolveId`?

**Review by vladfrangu — APPROVED** [review]

File: packages/discord.js/typings/index.d.ts (hunk 6 above)
Comment (Jiralite) [reply packages/discord.js/typings/index.d.ts:4075]: @almeidx any reason this wasn't done?

File: packages/discord.js/typings/index.d.ts [hunk 8]
```diff
@@ -3995,6 +4055,37 @@ export class ChannelManager extends CachedManager<Snowflake, Channel, ChannelRes
...
+  skus?: readonly SKUResolvable[];
+  excludeEnded?: boolean;
+  cache?: boolean;
+  before?: Snowflake;
+  after?: Snowflake;
+}
+
+export class EntitlementManager extends CachedManager<Snowflake, Entitlement, EntitlementResolvable> {
+  private constructor(client: Client<true>, iterable: Iterable<APIEntitlement>);
+  public fetch(options?: FetchEntitlementOptions): Promise<Collection<Snowflake, Entitlement>>;
```
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:4084]: `FetchEntitlementsOptions` or `FetchEntitlementOptions`? The JSDoc and the type differ here.

**Review by Jiralite — APPROVED** [review]

**Comment by Wuemeli** [comment]
How long will this take to be approved?

**Comment by didinele** [comment]
> How long will this take to be approved?

We don't do ETAs.

**Review by kyranet — APPROVED** [review]

**Pull request merged**